src/java/test/org/apache/zookeeper/server/HighLatencyRequestLoggingTest.java

```diff
@@ -0,0 +1,285 @@
+                        ClientBase.CONNECTION_TIMEOUT));
+
+
+    }
+
+    // Test to verify count of requests that exceeded threshold and logger rate limiting with a single thread
+    @Test
+    public void testSingleThreadFrequentRequestWarningThresholdLogging() throws IOException, KeeperException, InterruptedException {
+        ClientBase.setupTestEnv();
+
+        final int CLIENT_PORT = PortAssignment.unique();
+
+        ZooKeeperServerMainTest.MainThread main = new
+                ZooKeeperServerMainTest.MainThread(CLIENT_PORT, true, null, 0);
+        main.start();
+
+        Assert.assertTrue("waiting for server being up",
+                ClientBase.waitForServerUp("127.0.0.1:" + CLIENT_PORT,
+                        CONNECTION_TIMEOUT));
+        // Get the stats object from the ZooKeeperServer to keep track of high latency requests.
+        ServerStats stats = main.main.getServerStats();
+        stats.setWaitForLoggingWarnThresholdMsgToFalse();
+        stats.setDelayTimeForLoggingWarnThresholdMsg(1000);
+
+        ZooKeeper zk = new ZooKeeper("127.0.0.1:" + CLIENT_PORT,
+                ClientBase.CONNECTION_TIMEOUT, this);
+
+        Thread thread = new Thread(new ZNodeCreator(zk, 0, 5));
+        thread.start();
+        thread.join();
+
+        zk.close();
+        main.shutdown();
+        main.join();
+        main.deleteDirs();
+
+        // Total Requests: 12
+        // 1 Thread, each writing and verifying 5 znodes = (1 threads * 5 znodes * 2 req = 10)
+        // CreateSession, CloseSession
+        verifyLogMessages(12, 5);
+
+        Assert.assertTrue("waiting for server down",
+                ClientBase.waitForServerDown("127.0.0.1:" + CLIENT_PORT,
+                        ClientBase.CONNECTION_TIMEOUT));
+    }
+
+
+    // Test to verify count of requests that exceeded threshold and logger rate limiting with a multiple threads and multiple times
+    @Test
+    public void testMultipleThreadsFrequentRequestWarningThresholdLogging()
+            throws IOException, KeeperException, InterruptedException {
+        ClientBase.setupTestEnv();
+
+        final int CLIENT_PORT = PortAssignment.unique();
+
+        ZooKeeperServerMainTest.MainThread main =
+                new ZooKeeperServerMainTest.MainThread(CLIENT_PORT, true, null, 0);
+        main.start();
+
+        Assert.assertTrue("waiting for server being up",
+                ClientBase.waitForServerUp("127.0.0.1:" + CLIENT_PORT,
+                        CONNECTION_TIMEOUT));
+        // Get the stats object from the ZooKeeperServer to keep track of high latency requests.
+        ServerStats stats = main.main.getServerStats();
+        stats.setWaitForLoggingWarnThresholdMsgToFalse();
+        stats.setDelayTimeForLoggingWarnThresholdMsg(1000);
+
+        final ZooKeeper zk = new ZooKeeper("127.0.0.1:" + CLIENT_PORT,
+                ClientBase.CONNECTION_TIMEOUT, this);
+
+        Thread thread1 = new Thread(new ZNodeCreator(zk,0,3));
```

Comment:
Using a for loop to housekeep these threads would have the additional benefit of testing with more clients (like 100).